### ISSUE TITLE
proc_macro_span: assert2 doesn't like being wrapped in a macro on nightly

### ISSUE THREAD

**Issue by ComputerDruid** [de-vri-es/assert2-rs#13]
I was experimenting with assert2 by writing an `assert_eq!` macro that shadows the std one and forwards to assert2, so I could see how it would look on my existing tests, and came across the following example:

```rust
#[test]
fn test_assert2() {
    macro_rules! assert_eq {
        ($left:expr, $right:expr) => {
            ::assert2::assert!($left == $right)
        };
    }
    assert_eq!("foo".as_bytes(), "bar".as_bytes());
}
```

On stable (1.44.1), it produces:

```
Assertion failed at tests/assert2.rs:8:5:
  assert!( "foo" . as_bytes() == "bar" . as_bytes() )
with expansion:
  [102, 111, 111] == [98, 97, 114]
```

The formatting of the expression is relatively ugly, but it works as expected.

However, on nightly (1.46, 2020-07-01):

```
Assertion failed at tests/assert2.rs:8:5:
  assert!( #[proc_macro_hack]
pub use assert2_macros::check_impl #[proc_macro_hack]
pub use assert2_macros::check_impl #[proc_macro_hack]
pub use assert2_macros::check_impl )
with expansion:
  [102, 111, 111] #[proc_macro_hack]
pub use assert2_macros::check_impl [98, 97, 114]
```

The output is basically unusable.

**Comment by de-vri-es** [comment]
Interesting. Without proc-macro-hack the output on nightly is better:

```
Assertion failed at src/main.rs:12:2:
  assert!( $left == $right )
with expansion:
  [102, 111, 111] == [98, 97, 114]
```

But obviously `$left` and `$right` aren't ideal either.

Interestingly, if we change `$val:expr` to `$(val:tt)*`, the output is as expected. However, that requires putting parenthesis around the arguments as `($($left:tt)*, $($right:tt)*) => { ... }` is ambiguous. So not a real solution either.

The end result would be:
```rust
#[test]
fn test_assert2() {
    macro_rules! assert_eq {
        (($($left:tt)*), ($($right:tt)*)) => {
            ::assert2::assert!($($left)* == $($right)*)
        };
    }
    assert_eq!(("foo".as_bytes()), ("bar".as_bytes()));
}
```

With output:
```
Assertion failed at src/main.rs:12:2:
  assert!( "foo".as_bytes() == "bar".as_bytes() )
with expansion:
  [102, 111, 111] == [98, 97, 114]
```

At least expr and literal fragments give the problem you showed, probably everything except for tt fragments. Maybe this is a bug in the rust compiler. The stringification on nightly does use the `proc_macro_span` feature which is still unstable.

It is also unclear to me why this fixes the comparison operator. It is used as separate token in the macros, so I don't get why the fragment specifiers of `$left` and `$right` influence it.

A nasty solution that would probably work is to wrap it with a `proc_macro` instead of `macro_rules`. But that's way too much overhead for something simple.

**Comment by m-ou-se** [comment]
The problem with displaying the expanded form `assert!( "foo".as_bytes() == "bar".as_bytes() )` instead of `assert!( $left == $right )` as it appears in the source code is that the information about grouping is gone:

```rust
    macro_rules! assert_eq {
        ($left:expr, $right:expr) => {
            ::assert2::assert!($left == $right)
        };
    }
    assert_eq!(true, true && false);
```

would give:
![image](https://user-images.githubusercontent.com/783247/86512432-99564000-be02-11ea-876d-f15f279865b8.png)
but `true == true && false` would group as `(true == true) && false`, not `true == (true && false)`.

Maybe it should show something like:

```
Assertion failed at src/main.rs:12:2:
  assert!( $left = $right )
with:
  $left = "foo".as_bytes()
  $right = "bar".as_bytes()
with expansion:
  [102, 111, 111] == [98, 97, 114]
```

**Comment by m-ou-se** [comment]
The version on the `beta` branch is no longer using `proc_macro_hack` and now displays only `assert!( $left = $right )`. It requires rust 1.45, which will be stable in 12 days.

**Comment by m-ou-se** [comment]
Good news and bad news:

The good news is that expansion of the macro fragments is working now in a test branch:

![image](https://user-images.githubusercontent.com/783247/86515467-50f64c80-be19-11ea-94dd-4a08e27431ea.png)

The bad news is that we found out hygiene is broken, also on the released version. The assert should not have failed, as `(1 + 2) * (2 + 3) == 15`. Somehow, the implicit grouping of the macro fragments doesn't survive `syn`'s parsing, even though the correct `Group`s do appear in the resulting token stream. The expression is evaluated as `1 + 2 * 2 + 3`, which is C-preprocessor behaviour, and should definitely not happen in Rust.

**Comment by de-vri-es** [comment]
Note that as a side-effect of the fixes by @m-ou-se, the output on stable and beta will also be a bit nicer, as the expression will be stringified by `stringify!`, where `stringify!` knows it is an expression. That gives a better result than just stringifying the tokens from the `proc_macro`.

**Comment by m-ou-se** [comment]
`beta` branch is ready. Just waiting for Rust 1.45 to be released now. Should happen on the 16th.

**Comment by de-vri-es** [comment]
Applied some final fixes, added some tests, and it's now released as 0.3.0 \o/

**Comment by de-vri-es** [comment]
@ComputerDruid: Thanks for reporting the issue, and @m-ou-se thanks for the fixes :)
